src/main/java/io/strimzi/kafka/bridge/http/HttpSourceBridgeEndpoint.java

```diff
@@ -126,48 +124,52 @@ public void handle(Endpoint<?> endpoint, Handler<?> handler) {
-                    results.add(new HttpBridgeResult<>(metadata));
-                } else {
-                    String msg = sendHandlers.get(i).cause().getMessage();
-                    int code = handleError(sendHandlers.get(i).cause());
-                    log.error("Failed to deliver record {}", finalRecords.get(i), done.cause());
-                    results.add(new HttpBridgeResult<>(new HttpBridgeError(code, msg)));
-                }
+            @SuppressWarnings({ "rawtypes" })
+            List<CompletableFuture> promises = new ArrayList<>(records.size());
+            for (ProducerRecord<K, V> record : records) {
+                CompletionStage<RecordMetadata> sendHandler =
+                        // inside send method, the callback which completes the promise is executed in the kafka-producer-network-thread
+                        // let's do the result handling async to free the network thread for more sends in parallel
+                        this.send(record, false).handle((metadata, ex) -> {
+                            log.trace("Handle thread {}", Thread.currentThread());
+                            if (ex == null) {
+                                log.debug("Delivered record {} to Kafka on topic {} at partition {} [{}]", record, metadata.topic(), metadata.partition(), metadata.offset());
+                                results.add(new HttpBridgeResult<>(metadata));
+                            } else {
+                                String msg = ex.getMessage();
+                                int code = handleError(ex);
+                                log.error("Failed to deliver record {}", record, ex);
+                                results.add(new HttpBridgeResult<>(new HttpBridgeError(code, msg)));
+                            }
+                            return metadata;
+                        });
+                promises.add(sendHandler.toCompletableFuture());
             }
-            // always return OK, since failure cause is in the response, per message
-            span.finish(HttpResponseStatus.OK.code());
-            HttpUtils.sendResponse(routingContext, HttpResponseStatus.OK.code(),
-                BridgeContentType.KAFKA_JSON, buildOffsets(results).toBuffer());
-            this.maybeClose();
+
+            CompletableFuture.allOf(promises.toArray(new CompletableFuture[0]))
+                    .whenCompleteAsync((v, t) -> {
+                        log.trace("All sent thread {}", Thread.currentThread());
+                        // always return OK, since failure cause is in the response, per message
+                        span.finish(HttpResponseStatus.OK.code());
+                        HttpUtils.sendResponse(routingContext, HttpResponseStatus.OK.code(),
+                                BridgeContentType.KAFKA_JSON, buildOffsets(results).toBuffer());
```

Comment:
The thread safety of access to `results` was pretty unclear to me. AFAICS:

1. Results get `add`ed to it on the Producer's IO thread.
2. The `CompletableFuture.allOf` would also complete on the Producer's IO thread, 
3. But use of `whenCompleteAsync` means at the point we're wanting to use `results` we're on a thread that's provided by the "stage's default  asynchronous execution facility" (I _guess_ that's a thread from the common ForkJoinPool).
4. Therefore there's nothing providing synchronization between the adding of results to `results` and the use of `results` here.

Obviously we _could_ solve that by using a synchronized list, but I think part of the problem here is that there isn't really a documented model for how the threading is _supposed_ to work. 

VertX has a model: You access everything from a single thread (the event loop thread aka context thread) and so no synchronization is required. The Vertx Kafka client is careful to work with this model by completing futures/running handlers on the context thread, and not exposing mutable state between the worker threads and the context thread.

We could adopt the same model or do something else. But whatever we do we need to document how it's supposed to work.

Reply:
So ...

Let's first talk about the "thread safety of access to `results`" whatever is the thread where each step ends.
The `allOf` is anyway waiting that all the completable future ends, so that they have accessed to `results` to add more ... results. When this happens, there is no more other threads adding results so it can be accessed safely to be read for sending the HTTP response. It looks to me that the "thread safety" is guaranteed by the semantics of `allOf`: execute the `whenComplete` only when ALL the completable future in this list complete. Of course, suing a synchronized list doesn't hurt at all.

> Obviously we could solve that by using a synchronized list, but I think part of the problem here is that there isn't really a documented model for how the threading is supposed to work.

Regarding the threads ... I read a lot but it seems that how threading is handled when using completable futures is not well documented and sometimes unclear. On each completion stage you don't have the guarantee that the next one will be on the same thread, it could happen in a different one. So going back to have a stage to be executed in the thread of the previous stage (compared to Vert.x, calling the handler on the context thread and not worker thread) is not possible AFAICS. But again for the explanation above I don't see to be a problem here specifically.

Back to your points instead ...

1. That's right
2. Debugging it seems not true ... I see the `allOf` executed in a thread from the fork-join pool
3. When using the `whenComplete` instead, I see it's executed in the Producer IO thread again. My intention was to free that thread as soon as possible.

Reply:
> When this happens, there is no more other threads adding results so it can be accessed safely to be read for sending the HTTP response.

I think what matters is whether there's a _happens-before_ relationship between the action of results being added (in one thread) and the action of using the `results` (in some other thread).

>  It looks to me that the "thread safety" is guaranteed by the semantics of `allOf`

Yes, I think you're right. I was forgetting how this

> Actions taken by the asynchronous computation represented by a Future happen-before actions subsequent to the retrieval of the result via Future.get() in another thread. 

(which is from https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility, a place that's a whole lot less discoverable than, say, the Javadoc for `CompletableFuture`, where someone might reasonably look for it), must imply that (given some reasonable assumptions about how there must be a happens before on the `CF.get` and the execution of code in the async handler).

> On each completion stage you don't have the guarantee that the next one will be on the same thread, it could happen in a different one. 

Yes, that's true, but:

1. Via the _happens-before_ logic it usually doesn't matter
2. You'd typically only get the completion on a different thread if you used `whenCompleteAsync`. Plain old `whenComplete` will give you behaviour that's closer to the Vert.X semantics (which makes we wonder why you're using the Async variant here, actually). 

> 2. Debugging it seems not true ... I see the allOf executed in a thread from the fork-join pool

Interesting. I guess it's not _documented_ to execute on the thread on which the last CF was completed. 

> 3. When using the whenComplete instead, I see it's executed in the Producer IO thread again. My intention was to free that thread as soon as possible.

I agree that that's what we should be aiming for. Did you consider putting the logic in `SourceBridgeEndpoint.send`? I.e. in your producer `Callback` you can `submit` some runnable to an `Executor` (it could be the common ForkJoinPool if you're sticking with that) and that runnable is the thing to call `complete()`. Note that by using a single thread executor you can get back to something closer to the VertX model.

Reply:
I was trying to figure out what exactly we want to achieve here :-)
Let's consider the current thread model, by running an example with an HTTP client sending 3 records in one HTTP request. The log about the threads will be the following ...

```shell
*(1)* [2022-12-16 15:48:36,831] INFO  <idgeEndpoint:127> [oop-thread-1] Handle thread Thread[vert.x-eventloop-thread-1,5,main]
[2022-12-16 15:48:36,833] INFO  <send        :95> [oop-thread-1] [1560561002] SEND Response:  statusCode = 200, message = OK
*(2)* [2022-12-16 15:48:36,833] INFO  <idgeEndpoint:131> [ool-worker-1] Handle runAsync thread Thread[ForkJoinPool.commonPool-worker-1,5,main]
[2022-12-16 15:48:36,844] INFO  <Metadata    :402> [| producer-1] [Producer clientId=producer-1] Resetting the last seen epoch of partition my-topic-0 to 0 since the associated topicId changed from null to 8QyCRfxTTmunDvzyAsDS7A
*(3)* [2022-12-16 15:48:36,872] INFO  <idgeEndpoint:151> [| producer-1] Send handle thread Thread[kafka-producer-network-thread | producer-1,5,main]
*(3)* [2022-12-16 15:48:36,873] INFO  <idgeEndpoint:151> [| producer-1] Send handle thread Thread[kafka-producer-network-thread | producer-1,5,main]
*(3)* [2022-12-16 15:48:36,873] INFO  <idgeEndpoint:151> [| producer-1] Send handle thread Thread[kafka-producer-network-thread | producer-1,5,main]
*(4)* [2022-12-16 15:48:36,874] INFO  <idgeEndpoint:168> [ool-worker-1] All sent thread Thread[ForkJoinPool.commonPool-worker-1,5,main]
```

Following what's happening ...

(1) Entering the `HttpSourceBridgeEndpoint.handle` and we are in the Vert.x event loop thread that we have to free asap to avoid the usual warning messages
(2) This is where we are leaving the Vert.x event loop thread and running the overall sending code through a `CompletableFuture.runAsync` on line 129. We are in the `ForkJoinPool.commonPool-worker-1` from the forkjoin pool. The loop of sending the 3 records runs in this thread to guarantee the order.
(3) This is where the Kafka callback is called so the part where we fill the `results` list (from line 149).
(4) Finally when all sent are completed we sent back the HTTP response. Because of using `whenCompleteAsync`, we are in a different thread (even if in this case JVM is reusing the same `ForkJoinPool.commonPool-worker-1` (but it should be only a case not a rule).

Now, if we switch to use `whenComplete`, we said that the HTTP response will be sent from the Kafka producer thread, something we don't want to take it busy. 
Now, making your proposal by using a single thread executor in the Kafka producer callback would mean using that thread at point (3) and then we can "reuse" it at point (4) by changing to `whenComplete`.

My question is what do you think is wrong with the current threads flow?
Is it just about freeing the Kafka producer thread faster than now, so in the callback?
Or is there anything more you want to achieve here?

Reply:
> Or is there anything more you want to achieve here?

I'd like to have something that's a) understandable and b) commented sufficiently for other contributors.

Reply:
While I can help on b) by writing more comments, what is not making a) ? Just comments missing? So it a) a consequence of b) ... otherwise I am still lost on what you mean :-)